// @blockfrost/blockfrost-js/lib/utils/errors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.handleError = exports.isBlockfrostErrorResponse = exports.BlockfrostClientError = exports.BlockfrostServerError = void 0;
const got_1 = require("got");
class BlockfrostServerError extends Error {
    constructor(error) {
        super(error.message);
        this.name = 'BlockfrostServerError';
        this.status_code = error.status_code;
        this.message = error.message;
        this.error = error.error;
        Object.setPrototypeOf(this, BlockfrostServerError.prototype);
    }
}
exports.BlockfrostServerError = BlockfrostServerError;
class BlockfrostClientError extends Error {
    constructor(error) {
        super(error.message);
        this.name = 'BlockfrostClientError';
        this.code = error.code;
        this.message = error.message;
        Object.setPrototypeOf(this, BlockfrostClientError.prototype);
    }
}
exports.BlockfrostClientError = BlockfrostClientError;
const hasProp = (
// eslint-disable-next-line @typescript-eslint/ban-types
data, prop) => {
    return prop in data;
};
const isBlockfrostErrorResponse = (data) => {
    // type guard for narrowing response body to an error object that should be returned by Blockfrost API
    return (typeof data === 'object' &&
        data !== null &&
        hasProp(data, 'status_code') &&
        hasProp(data, 'message') &&
        hasProp(data, 'error'));
};
exports.isBlockfrostErrorResponse = isBlockfrostErrorResponse;
const handleError = (error) => {
    var _a, _b;
    if (error instanceof got_1.HTTPError) {
        const responseBody = error.response.body;
        if ((0, exports.isBlockfrostErrorResponse)(responseBody)) {
            return new BlockfrostServerError(responseBody);
        }
        else {
            // response.body may contain html output (eg. errors returned by nginx)
            const { statusCode } = error.response;
            const statusText = (_a = error.response.statusMessage) !== null && _a !== void 0 ? _a : error.message;
            return new BlockfrostServerError({
                status_code: statusCode,
                message: `${statusCode}: ${statusText}`,
                error: statusText,
            });
        }
    }
    // system errors such as -3008 ENOTFOUND and various got errors like ReadError, CacheError, MaxRedirectsError, TimeoutError,...
    // https://github.com/sindresorhus/got/blob/main/documentation/8-errors.md
    return new BlockfrostClientError({
        code: (_b = error.code) !== null && _b !== void 0 ? _b : 'ERR_GOT_REQUEST_ERROR',
        message: error.message, // getaddrinfo ENOTFOUND cardano-testnet.blockfrost.io'
    });
};
exports.handleError = handleError;
